### PR TITLE
[User] start exam

### PR DESCRIPTION
**Pull request by kagoromo**

### CODE REVIEW

File: config/routes.rb [hunk 1]
```diff
@@ -33,6 +33,6 @@
 
   namespace :users do
     root "exams#index"
-    resources :exams, only: [:index, :create]
+    resources :exams, only: [:index, :show, :create, :update]
```
Comment (hatd) [inline config/routes.rb]: dufng except

File: app/models/exam.rb [hunk 2]
```diff
@@ -15,6 +16,25 @@ def calculated_score
...
+    time = interval > subject.duration * 60 ? subject.duration * 60 : interval
+  end
+
+  def spent_time_format
+    Time.at(spent_time).utc.strftime("%H:%M:%S")
+  end
+
+  def time_out?
+    unchecked? || checked? ||
+      (Time.zone.now > (started_at.nil? ? updated_at : started_at) + subject.duration.minutes)
```
Comment (tranduytung) [inline app/models/exam.rb]: Quá 80 kí tự

File: app/views/users/exams/show.html.erb [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+    <%= l @exam.started_at, format: :short %></br>
+    <% if @exam.checked? %>
+      <%= t "exam.result" %>
+      <span><%= @exam.score %></span>
+    <% end %>
+    <div class="text-center">
+      <h2><%= @subject.content %></h2>
+    </div>
+    <hr>
+    <% if @exam.time_out? %>
```
Comment (tranduytung) [inline app/views/users/exams/show.html.erb]: refactor code đi

File: app/views/users/exams/_exam_form.html.erb [hunk 4]
```diff
@@ -0,0 +1,29 @@
+<%= form_for [:users, @exam] do |f| %>
+  <ol>
+    <%= f.fields_for :results do |ff| %>
+      <li><%= ff.object.question.content %></li>
+      <br/>
+      <% case ff.object.question.question_type %>
+      <% when Settings.questions.multiple_choice %>
+        <%= ff.collection_check_boxes :answer_ids, ff.object.question.answers,
+          :id, :content, {}, {disabled: state} do |cb| %>
+            <%= cb.label(class: "check_box") {cb.check_box(class: "cb") + cb.text} %>
```
Comment (tranduytung) [inline app/views/users/exams/_exam_form.html.erb]: Quá 80 kí tự

**Comment by tranduytung** [comment]
OK

File: app/controllers/users/exams_controller.rb [hunk 5]
```diff
@@ -1,6 +1,6 @@
 class Users::ExamsController < ApplicationController
   before_action :authenticate_user!
-  authorize_resource
+  load_and_authorize_resource except: :show
```
Comment (ngtrduc) [inline app/controllers/users/exams_controller.rb]: skip_load thoi, con van phai authorize

File: app/models/exam.rb [hunk 6]
```diff
@@ -15,6 +16,25 @@ def calculated_score
     results.correct.count
   end
 
+  def calculated_spent_time
+    interval = Time.zone.now - started_at
+    time = interval > subject.duration * 60 ? subject.duration * 60 : interval
+  end
+
+  def spent_time_format
+    Time.at(spent_time).utc.strftime("%H:%M:%S")
```
Comment (ngtrduc) [inline app/models/exam.rb]: I18n

File: app/views/users/exams/_search_form.html.erb [hunk 7]
```diff
@@ -10,7 +10,7 @@
     <div class="col-md-2">
       <%= t "search.status" %>
       <%= f.select :status_eq, @statuses.collect{|key, value| [t(key), value]},
-        {include_blank: t("all")} %>
+        {include_blank: t("all")}, class: "form-control" %>
```
Comment (ngtrduc) [inline app/views/users/exams/_search_form.html.erb]: bo ()

File: app/controllers/users/exams_controller.rb [hunk 8]
```diff
@@ -13,16 +14,44 @@ def index
       order created_at: :desc
   end
 
+  def show
+    @exam = Exam.includes({results: [{question: :answers}, :text_answer]}).
+      find_by(id: params[:id])
```
Comment (hellovietnam93) [inline app/controllers/users/exams_controller.rb]: bỏ ()

File: app/controllers/users/exams_controller.rb [hunk 9]
```diff
@@ -13,16 +14,44 @@ def index
...
-      flash[:danger] = t "exam.not_created"
+      flash[:danger] = t "exam.subject_not_found"
     end
     redirect_to users_exams_path
   end
 
+  def update
+    if @exam.update_attributes exam_params
+      @exam.update_attributes spent_time: @exam.calculated_spent_time
+      if @exam.time_out?
```
Comment (hellovietnam93) [inline app/controllers/users/exams_controller.rb]: exam.unchecked! if @exam.time_out?

File: app/models/result.rb [hunk 10]
```diff
@@ -6,4 +6,12 @@ class Result < ActiveRecord::Base
   has_one :text_answer, dependent: :destroy
 
   scope :correct, ->{where is_correct: true}
+
+  accepts_nested_attributes_for :text_answer
+
+  after_create :create_text_answer
+
+  def create_text_answer
```
Comment (hellovietnam93) [inline app/models/result.rb]: private method ?

File: app/controllers/users/exams_controller.rb [hunk 11]
```diff
@@ -1,5 +1,6 @@
 class Users::ExamsController < ApplicationController
   before_action :authenticate_user!
+  load_resource except: :show
```
Comment (hellovietnam93) [inline app/controllers/users/exams_controller.rb]: load trước authorize à?

**Pull request merged**